File: src/pytra/std/sys.py
"""pytra.std.sys: extern-marked sys API with Python runtime fallback."""

from __future__ import annotations

from pytra.std import extern

import sys as __s

argv: list[str] = extern(__s.argv)
path: list[str] = extern(__s.path)
stderr: object = extern(__s.stderr)
stdout: object = extern(__s.stdout)


@extern
def exit(code: int = 0) -> None:
    __s.exit(code)


@extern
def set_argv(values: list[str]) -> None:
    argv.clear()
    for v in values:
        argv.append(v)


@extern
def set_path(values: list[str]) -> None:
    path.clear()
    for v in values:
        path.append(v)


@extern
def write_stderr(text: str) -> None:
    __s.stderr.write(text)


@extern
def write_stdout(text: str) -> None:
    __s.stdout.write(text)

File: src/runtime/cpp/std/sys.cpp
#include "core/py_runtime.h"
#include "core/process_runtime.h"

#include "std/sys.h"

list<str> argv = py_runtime_argv();
list<str> path = {};

void exit(int64 code) {
    py_runtime_exit(code);
}

void set_argv(const list<str>& values) {
    argv = values;
    py_runtime_set_argv(values);
}

void set_path(const list<str>& values) {
    path = values;
}

void write_stderr(const str& text) {
    py_runtime_write_stderr(text);
}

void write_stdout(const str& text) {
    py_runtime_write_stdout(text);
}
